blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,149 @@
+
+## A deeper contributing factor
+
+Over time, our engineering organization formed some tribal knowledge regarding ways to avoid "taking down production" with poorly defined migrations. Of particular hesitation was creating indexes on tables that can have a large number of rows. Normal index creation operations take an exclusive write lock on the table, creating the new index from the existing table data synchronously. During this time no inserts, updates, or deletes can be made to the table. You basically incur write downtime (along with a blip of failed requests in your metrics dashboards and alerting systems, and a slice of failed jobs that failed in the background) for the table, which can take tens of minutes for large tables.
+
+For our use case, `CREATE INDEX CONCURRENTLY` was received as a hero.
+
+Using concurrent index creation reduced downtime incidents in our Cloud environment related to heavy database contention triggered by continuous deployment.
+
+Using concurrent index creation also introduced a footgun and plenty of friendly fire that we didn't recognize until recently. [Concurrent index creation](https://www.postgresql.org/docs/12/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY) is performed in three distinct phases to ensure that a full table lock is never required, thus readers nor writers are not blocked for long stretches of time.
+
+- **Phase 1**: Create new catalog structures for the index, but do not begin to populate them yet. Transactions that started before the completion of this phase will only see (and therefore only write to) the old catalog structures.
+- **Phase 2**: Wait for all transactions starting before the end of **Phase 1** to finish. Then, perform a full-table scan and write the content of each row to the new index structure. Open the new index structure up to writes.
+- **Phase 3**: Wait for all transactions starting before the end of **Phase 2** to finish. Then, perform another full-table scan and write the content of each row inserted during **Phase 2** to the new index structure. The old index structure can now be closed to writes.
+
+There's a particular detail here that deserves special attention in hindsight: concurrent index creation operations block until **all transactions started before them** have either committed or rolled back. Mixed with long-held advisory locks, we encounter this rather nasty situation.
+
+Observe! In `psql` **Process A**, we acquire an advisory lock:
+
+```sql
+[A] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
+ pg_advisory_lock
+------------------
+
+(1 row)
+
+Time: 20.616 ms
+```
+
+In `psql` **Process B**, we attempt to acquire an advisory lock:
+
+```sql
+[B] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
+
+```
+
+As this lock is already held by another process, this command blocks indefinitely - or until the user gets bored and aborts their session.
```

Comment:
```suggestion
Because this lock is already held by another process, this command blocks indefinitely–or until the user gets bored and aborts their session.
```